File: packages/naver-map/ios/BuiltInNaverMapProvider.swift
//
//  BuiltInNaverMapProvider.swift
//  granite-naver-map
//
//  Built-in provider using NMapsMap SDK
//

import UIKit
import NMapsMap

@objc public class BuiltInNaverMapProvider: NSObject, NaverMapProvidable {
    private var mapView: NMFNaverMapView?
    private weak var delegate: NaverMapProviderDelegate?
    private var isInitialized = false

    private var markers: [String: NMFMarker] = [:]
    private var polylines: [String: NMFPolylineOverlay] = [:]
    private var polygons: [String: NMFPolygonOverlay] = [:]
    private var circles: [String: NMFCircleOverlay] = [:]
    private var paths: [String: NMFPath] = [:]

    // MARK: - NaverMapProvidable

    @objc public func createMapView(frame: CGRect) -> UIView {
        let view = NMFNaverMapView(frame: frame)
        self.mapView = view
        setupMap(view)
        return view
    }

    @objc public func setDelegate(_ delegate: NaverMapProviderDelegate?) {
        self.delegate = delegate
    }

    private func setupMap(_ view: NMFNaverMapView) {
        view.mapView.touchDelegate = self
        view.mapView.addCameraDelegate(delegate: self)
    }

    // MARK: - Camera

    @objc public func moveCamera(to position: NaverMapCameraPosition, animated: Bool) {
        guard let mapView = mapView else { return }
        let cameraPosition = NMFCameraPosition(
            NMGLatLng(lat: position.target.latitude, lng: position.target.longitude),
            zoom: position.zoom,
            tilt: position.tilt,
            heading: position.bearing
        )
        let update = NMFCameraUpdate(position: cameraPosition)
        if animated {
            update.animation = .easeIn
        }
        mapView.mapView.moveCamera(update)
    }

    @objc public func animateToCoordinate(_ coordinate: NaverMapCoordinate) {
        guard let mapView = mapView else { return }
        let cameraUpdate = NMFCameraUpdate(scrollTo: NMGLatLng(lat: coordinate.latitude, lng: coordinate.longitude))
        cameraUpdate.animation = .easeIn
        mapView.mapView.moveCamera(cameraUpdate)
    }

    @objc public func animateToBounds(_ bounds: NaverMapBounds, padding: CGFloat) {
        guard let mapView = mapView else { return }
        let nmgBounds = NMGLatLngBounds(
            southWest: NMGLatLng(lat: bounds.southWest.latitude, lng: bounds.southWest.longitude),
            northEast: NMGLatLng(lat: bounds.northEast.latitude, lng: bounds.northEast.longitude)
        )
        let cameraUpdate = NMFCameraUpdate(fit: nmgBounds, padding: padding)
        cameraUpdate.animation = .easeIn
        mapView.mapView.moveCamera(cameraUpdate)
    }

    // MARK: - Map Properties

    @objc public func setMapType(_ type: NaverMapType) {
        guard let mapView = mapView, let nmfType = NMFMapType(rawValue: type.rawValue) else { return }
        mapView.mapView.mapType = nmfType
    }

    @objc public func setMapPadding(_ padding: UIEdgeInsets) {
        mapView?.mapView.contentInset = padding
    }

    @objc public func setCompassEnabled(_ enabled: Bool) {
        mapView?.showCompass = enabled
    }

    @objc public func setScaleBarEnabled(_ enabled: Bool) {
        mapView?.showScaleBar = enabled
    }

    @objc public func setZoomControlEnabled(_ enabled: Bool) {
        mapView?.showZoomControls = enabled
    }

    @objc public func setLocationButtonEnabled(_ enabled: Bool) {
        mapView?.showLocationButton = enabled
    }

    @objc public func setBuildingHeight(_ height: Float) {
        mapView?.mapView.buildingHeight = height
    }

    @objc public func setNightModeEnabled(_ enabled: Bool) {
        mapView?.mapView.isNightModeEnabled = enabled
    }

    @objc public func setMinZoomLevel(_ level: Double) {
        mapView?.mapView.minZoomLevel = level
    }

    @objc public func setMaxZoomLevel(_ level: Double) {
        mapView?.mapView.maxZoomLevel = level
    }

    @objc public func setScrollGesturesEnabled(_ enabled: Bool) {
        mapView?.mapView.isScrollGestureEnabled = enabled
    }

    @objc public func setZoomGesturesEnabled(_ enabled: Bool) {
        mapView?.mapView.isZoomGestureEnabled = enabled
    }

    @objc public func setTiltGesturesEnabled(_ enabled: Bool) {
        mapView?.mapView.isTiltGestureEnabled = enabled
    }

    @objc public func setRotateGesturesEnabled(_ enabled: Bool) {
        mapView?.mapView.isRotateGestureEnabled = enabled
    }

    @objc public func setStopGesturesEnabled(_ enabled: Bool) {
        mapView?.mapView.isStopGestureEnabled = enabled
    }

    @objc public func setLocationTrackingMode(_ mode: NaverMapLocationTrackingMode) {
        guard let mapView = mapView else { return }
        mapView.mapView.positionMode = NMFMyPositionMode(rawValue: UInt(mode.rawValue)) ?? .disabled
    }

    @objc public func setLayerGroupEnabled(group: String, enabled: Bool) {
        guard let mapView = mapView else { return }
        let layerGroups: [String: String] = [
            "building": NMF_LAYER_GROUP_BUILDING,
            "ctt": NMF_LAYER_GROUP_TRAFFIC,
            "transit": NMF_LAYER_GROUP_TRANSIT,
            "bike": NMF_LAYER_GROUP_BICYCLE,
            "mountain": NMF_LAYER_GROUP_MOUNTAIN,
            "landparcel": NMF_LAYER_GROUP_CADASTRAL
        ]
        guard let layerGroup = layerGroups[group] else { return }
        mapView.mapView.setLayerGroup(layerGroup, isEnabled: enabled)
    }

    // MARK: - Markers

    @objc public func addMarker(_ data: ProviderMarkerData) {
        guard let mapView = mapView else { return }

        let marker = NMFMarker()
        marker.position = NMGLatLng(lat: data.coordinate.latitude, lng: data.coordinate.longitude)
        marker.width = CGFloat(data.width)
        marker.height = CGFloat(data.height)
        marker.zIndex = data.zIndex
        marker.angle = CGFloat(data.rotation)
        marker.isFlat = data.flat
        marker.alpha = CGFloat(data.alpha)

        marker.mapView = mapView.mapView
        markers[data.identifier] = marker

        let identifier = data.identifier
        marker.touchHandler = { [weak self] _ -> Bool in
            self?.delegate?.mapViewDidClickMarker(id: identifier)
            return true
        }

        if !data.image.isEmpty {
            loadMarkerImage(marker: marker, image: data.image)
        }
    }

    @objc public func updateMarker(_ data: ProviderMarkerData) {
        guard let marker = markers[data.identifier] else { return }

        marker.position = NMGLatLng(lat: data.coordinate.latitude, lng: data.coordinate.longitude)
        marker.width = CGFloat(data.width)
        marker.height = CGFloat(data.height)
        marker.zIndex = data.zIndex
        marker.angle = CGFloat(data.rotation)
        marker.isFlat = data.flat
        marker.alpha = CGFloat(data.alpha)

        if !data.image.isEmpty {
            loadMarkerImage(marker: marker, image: data.image)
        }
    }

    @objc public func removeMarker(identifier: String) {
        guard let marker = markers[identifier] else { return }
        marker.mapView = nil
        markers[identifier] = nil
    }

    private func loadMarkerImage(marker: NMFMarker, image: String) {
        if image.hasPrefix("http://") || image.hasPrefix("https://") {
            guard let url = URL(string: image) else { return }
            Task {
                guard let (data, _) = try? await URLSession.shared.data(from: url),
                      let uiImage = UIImage(data: data) else { return }
                await MainActor.run {
                    marker.iconImage = NMFOverlayImage(image: uiImage)
                }
            }
        } else {
            marker.iconImage = NMFOverlayImage(name: image)
        }
    }

    // MARK: - Polylines

    private func colorFromInt(_ value: Int) -> UIColor {
        let uValue = UInt32(bitPattern: Int32(truncatingIfNeeded: value))
        let alpha = CGFloat((uValue >> 24) & 0xFF) / 255.0
        let red = CGFloat((uValue >> 16) & 0xFF) / 255.0
        let green = CGFloat((uValue >> 8) & 0xFF) / 255.0
        let blue = CGFloat(uValue & 0xFF) / 255.0
        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
    }

    private func lineCapType(_ value: Int) -> NMFOverlayLineCap {
        switch value {
        case 1: return .round
        case 2: return .square
        default: return .butt
        }
    }

    private func lineJoinType(_ value: Int) -> NMFOverlayLineJoin {
        switch value {
        case 1: return .miter
        case 2: return .round
        default: return .bevel
        }
    }

    @objc public func addPolyline(_ data: ProviderPolylineData) {
        guard let mapView = mapView else { return }

        let coords = data.coordinates.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) }
        guard coords.count >= 2 else { return }

        let polyline = NMFPolylineOverlay(NMGLineString(points: coords))
        polyline?.width = CGFloat(data.strokeWidth)
        polyline?.color = colorFromInt(data.strokeColor)
        polyline?.zIndex = data.zIndex
        polyline?.capType = lineCapType(data.lineCap)
        polyline?.joinType = lineJoinType(data.lineJoin)

        if !data.pattern.isEmpty {
            polyline?.pattern = data.pattern.map { NSNumber(value: $0) }
        }

        polyline?.mapView = mapView.mapView
        if let polyline = polyline {
            polylines[data.identifier] = polyline
        }
    }

    @objc public func updatePolyline(_ data: ProviderPolylineData) {
        guard let polyline = polylines[data.identifier] else { return }

        let coords = data.coordinates.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) }
        if coords.count >= 2 {
            polyline.line = NMGLineString(points: coords)
        }
        polyline.width = CGFloat(data.strokeWidth)
        polyline.color = colorFromInt(data.strokeColor)
        polyline.zIndex = data.zIndex
        polyline.capType = lineCapType(data.lineCap)
        polyline.joinType = lineJoinType(data.lineJoin)

        if !data.pattern.isEmpty {
            polyline.pattern = data.pattern.map { NSNumber(value: $0) }
        }
    }

    @objc public func removePolyline(identifier: String) {
        guard let polyline = polylines[identifier] else { return }
        polyline.mapView = nil
        polylines[identifier] = nil
    }

    // MARK: - Polygons

    @objc public func addPolygon(_ data: ProviderPolygonData) {
        guard let mapView = mapView else { return }

        let coords = data.coordinates.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) }
        guard coords.count >= 3 else { return }

        let exteriorRing = NMGLineString(points: coords)

        let nmgPolygon: NMGPolygon<AnyObject>
        if data.holes.isEmpty {
            nmgPolygon = unsafeBitCast(NMGPolygon(ring: exteriorRing), to: NMGPolygon<AnyObject>.self)
        } else {
            let interiorRings = data.holes.map { hole in
                NMGLineString(points: hole.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) })
            }
            nmgPolygon = unsafeBitCast(NMGPolygon(ring: exteriorRing, interiorRings: interiorRings), to: NMGPolygon<AnyObject>.self)
        }

        let polygon = NMFPolygonOverlay(nmgPolygon)
        polygon?.fillColor = colorFromInt(data.fillColor)
        polygon?.outlineColor = colorFromInt(data.strokeColor)
        polygon?.outlineWidth = UInt(data.strokeWidth)
        polygon?.zIndex = data.zIndex
        polygon?.mapView = mapView.mapView
        if let polygon = polygon {
            polygons[data.identifier] = polygon
        }
    }

    @objc public func updatePolygon(_ data: ProviderPolygonData) {
        guard let polygon = polygons[data.identifier] else { return }

        let coords = data.coordinates.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) }
        if coords.count >= 3 {
            let exteriorRing = NMGLineString(points: coords)

            if data.holes.isEmpty {
                let nmgPolygon = NMGPolygon(ring: exteriorRing)
                polygon.polygon = unsafeBitCast(nmgPolygon, to: NMGPolygon<AnyObject>.self)
            } else {
                let interiorRings = data.holes.map { hole in
                    NMGLineString(points: hole.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) })
                }
                let nmgPolygon = NMGPolygon(ring: exteriorRing, interiorRings: interiorRings)
                polygon.polygon = unsafeBitCast(nmgPolygon, to: NMGPolygon<AnyObject>.self)
            }
        }

        polygon.fillColor = colorFromInt(data.fillColor)
        polygon.outlineColor = colorFromInt(data.strokeColor)
        polygon.outlineWidth = UInt(data.strokeWidth)
        polygon.zIndex = data.zIndex
    }

    @objc public func removePolygon(identifier: String) {
        guard let polygon = polygons[identifier] else { return }
        polygon.mapView = nil
        polygons[identifier] = nil
    }

    // MARK: - Circles

    @objc public func addCircle(_ data: ProviderCircleData) {
        guard let mapView = mapView else { return }

        let circle = NMFCircleOverlay(NMGLatLng(lat: data.center.latitude, lng: data.center.longitude), radius: data.radius)
        circle.fillColor = colorFromInt(data.fillColor)
        circle.outlineColor = colorFromInt(data.strokeColor)
        circle.outlineWidth = Double(data.strokeWidth)
        circle.zIndex = data.zIndex
        circle.mapView = mapView.mapView
        circles[data.identifier] = circle
    }

    @objc public func updateCircle(_ data: ProviderCircleData) {
        guard let circle = circles[data.identifier] else { return }

        circle.center = NMGLatLng(lat: data.center.latitude, lng: data.center.longitude)
        circle.radius = data.radius
        circle.fillColor = colorFromInt(data.fillColor)
        circle.outlineColor = colorFromInt(data.strokeColor)
        circle.outlineWidth = Double(data.strokeWidth)
        circle.zIndex = data.zIndex
    }

    @objc public func removeCircle(identifier: String) {
        guard let circle = circles[identifier] else { return }
        circle.mapView = nil
        circles[identifier] = nil
    }

    // MARK: - Paths

    @objc public func addPath(_ data: ProviderPathData) {
        guard let mapView = mapView else { return }

        let coords = data.coordinates.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) }
        guard coords.count >= 2 else { return }

        let path = NMFPath(points: coords)
        path?.width = CGFloat(data.width)
        path?.outlineWidth = CGFloat(data.outlineWidth)
        path?.color = colorFromInt(data.color)
        path?.outlineColor = colorFromInt(data.outlineColor)
        path?.passedColor = colorFromInt(data.passedColor)
        path?.passedOutlineColor = colorFromInt(data.passedOutlineColor)
        path?.progress = Double(data.progress)
        path?.zIndex = data.zIndex

        if !data.patternImage.isEmpty {
            path?.patternIcon = NMFOverlayImage(name: data.patternImage)
            path?.patternInterval = UInt(data.patternInterval)
        }

        path?.mapView = mapView.mapView
        if let path = path {
            paths[data.identifier] = path
        }
    }

    @objc public func updatePath(_ data: ProviderPathData) {
        guard let path = paths[data.identifier] else { return }

        let coords = data.coordinates.map { NMGLatLng(lat: $0.latitude, lng: $0.longitude) }
        if coords.count >= 2 {
            path.path = NMGLineString(points: coords)
        }
        path.width = CGFloat(data.width)
        path.outlineWidth = CGFloat(data.outlineWidth)
        path.color = colorFromInt(data.color)
        path.outlineColor = colorFromInt(data.outlineColor)
        path.passedColor = colorFromInt(data.passedColor)
        path.passedOutlineColor = colorFromInt(data.passedOutlineColor)
        path.progress = Double(data.progress)
        path.zIndex = data.zIndex

        if !data.patternImage.isEmpty {
            path.patternIcon = NMFOverlayImage(name: data.patternImage)
            path.patternInterval = UInt(data.patternInterval)
        }
    }

    @objc public func removePath(identifier: String) {
        guard let path = paths[identifier] else { return }
        path.mapView = nil
        paths[identifier] = nil
    }

    // MARK: - Internal

    func notifyInitialized() {
        if !isInitialized {
            isInitialized = true
            delegate?.mapViewDidInitialize()
        }
    }
}

// MARK: - NMFMapViewTouchDelegate

extension BuiltInNaverMapProvider: NMFMapViewTouchDelegate {
    public func mapView(_ mapView: NMFMapView, didTapMap latlng: NMGLatLng, point: CGPoint) {
        delegate?.mapViewDidClick(x: point.x, y: point.y, latitude: latlng.lat, longitude: latlng.lng)
    }
}

// MARK: - NMFMapViewCameraDelegate

extension BuiltInNaverMapProvider: NMFMapViewCameraDelegate {
    public func mapView(_ mapView: NMFMapView, cameraWillChangeByReason reason: Int, animated: Bool) {
        delegate?.mapViewDidTouch(reason: reason, animated: animated)
    }

    public func mapViewCameraIdle(_ mapView: NMFMapView) {
        let position = mapView.cameraPosition
        let cameraPosition = NaverMapCameraPosition(
            target: NaverMapCoordinate(latitude: position.target.lat, longitude: position.target.lng),
            zoom: position.zoom,
            tilt: position.tilt,
            bearing: position.heading
        )
        delegate?.mapViewDidChangeCamera(position: cameraPosition)
    }
}

File: packages/naver-map/android/src/main/java/run/granite/navermap/BuiltInNaverMapProvider.kt
package run.granite.navermap

import android.content.Context
import android.graphics.BitmapFactory
import android.view.View
import android.widget.FrameLayout
import com.naver.maps.geometry.LatLng
import com.naver.maps.geometry.LatLngBounds
import com.naver.maps.map.CameraAnimation
import com.naver.maps.map.CameraPosition
import com.naver.maps.map.CameraUpdate
import com.naver.maps.map.LocationTrackingMode
import com.naver.maps.map.MapView
import com.naver.maps.map.NaverMap
import com.naver.maps.map.NaverMapOptions
import com.naver.maps.map.overlay.Marker
import com.naver.maps.map.overlay.OverlayImage
import com.naver.maps.map.overlay.PolylineOverlay
import com.naver.maps.map.overlay.PolygonOverlay
import com.naver.maps.map.overlay.CircleOverlay
import com.naver.maps.map.overlay.PathOverlay
import java.net.URL
import kotlin.concurrent.thread

/**
 * Built-in provider using NMapsMap SDK
 */
class BuiltInNaverMapProvider(private val context: Context) : NaverMapProvider {
    private var containerView: FrameLayout? = null
    private var mapView: MapView? = null
    private var naverMap: NaverMap? = null
    private var delegate: NaverMapProviderDelegate? = null
    private var mapReady = false
    private var mapInitialized = false

    private val density = context.resources.displayMetrics.density

    private val markers: MutableMap<String, Marker> = mutableMapOf()
    private val polylines: MutableMap<String, PolylineOverlay> = mutableMapOf()
    private val polygons: MutableMap<String, PolygonOverlay> = mutableMapOf()
    private val circles: MutableMap<String, CircleOverlay> = mutableMapOf()
    private val paths: MutableMap<String, PathOverlay> = mutableMapOf()

    private fun dpToPx(dp: Int): Int = (dp * density).toInt()
    private fun dpToPx(dp: Float): Int = (dp * density).toInt()

    override fun createMapView(context: Context): View {
        val container = FrameLayout(context)
        containerView = container

        val map = MapView(context, NaverMapOptions())
        mapView = map
        container.addView(map, FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        ))

        return container
    }

    override fun setDelegate(delegate: NaverMapProviderDelegate?) {
        this.delegate = delegate
    }

    override fun onAttachedToWindow() {
        // Lifecycle is handled in onSizeChanged
    }

    override fun onDetachedFromWindow() {
        mapView?.onPause()
        mapView?.onStop()
        mapView?.onDestroy()
    }

    override fun onSizeChanged(width: Int, height: Int) {
        if (!mapInitialized && width > 0 && height > 0) {
            mapInitialized = true
            initializeMap()
        }
    }

    override fun onHostResume() {
        if (mapReady) {
            mapView?.onResume()
        }
    }

    override fun onHostPause() {
        if (mapReady) {
            mapView?.onPause()
        }
    }

    private fun initializeMap() {
        mapView?.onCreate(null)
        mapView?.onStart()
        mapView?.onResume()

        mapView?.getMapAsync { map ->
            naverMap = map
            mapReady = true
            setupMapListeners(map)
            delegate?.onMapInitialized()
        }
    }

    private fun setupMapListeners(map: NaverMap) {
        map.addOnCameraIdleListener {
            val position = map.cameraPosition
            val contentBounds = map.contentBounds
            val coveringBounds = map.coveringBounds

            val cameraPos = NaverMapCameraPosition(
                target = NaverMapCoordinate(position.target.latitude, position.target.longitude),
                zoom = position.zoom,
                tilt = position.tilt,
                bearing = position.bearing
            )

            val contentRegion = listOf(
                NaverMapCoordinate(contentBounds.southLatitude, contentBounds.westLongitude),
                NaverMapCoordinate(contentBounds.southLatitude, contentBounds.eastLongitude),
                NaverMapCoordinate(contentBounds.northLatitude, contentBounds.eastLongitude),
                NaverMapCoordinate(contentBounds.northLatitude, contentBounds.westLongitude)
            )

            val coveringRegion = listOf(
                NaverMapCoordinate(coveringBounds.southLatitude, coveringBounds.westLongitude),
                NaverMapCoordinate(coveringBounds.southLatitude, coveringBounds.eastLongitude),
                NaverMapCoordinate(coveringBounds.northLatitude, coveringBounds.eastLongitude),
                NaverMapCoordinate(coveringBounds.northLatitude, coveringBounds.westLongitude)
            )

            delegate?.onCameraChange(cameraPos, contentRegion, coveringRegion)
        }

        map.addOnCameraChangeListener { reason, animated ->
            delegate?.onTouch(reason, animated)
        }

        map.setOnMapClickListener { point, latLng ->
            delegate?.onClick(point.x.toDouble(), point.y.toDouble(), latLng.latitude, latLng.longitude)
        }
    }

    // MARK: - Camera

    override fun moveCamera(position: NaverMapCameraPosition, animated: Boolean) {
        val cameraPosition = CameraPosition(
            LatLng(position.target.latitude, position.target.longitude),
            position.zoom,
            position.tilt,
            position.bearing
        )
        val update = CameraUpdate.toCameraPosition(cameraPosition)
        if (animated) {
            update.animate(CameraAnimation.Easing)
        }
        naverMap?.moveCamera(update)
    }

    override fun animateToCoordinate(coordinate: NaverMapCoordinate) {
        naverMap?.moveCamera(
            CameraUpdate.scrollTo(LatLng(coordinate.latitude, coordinate.longitude))
                .animate(CameraAnimation.Easing)
        )
    }

    override fun animateToBounds(bounds: NaverMapBounds, padding: Int) {
        val latLngBounds = LatLngBounds(
            LatLng(bounds.southWest.latitude, bounds.southWest.longitude),
            LatLng(bounds.northEast.latitude, bounds.northEast.longitude)
        )
        naverMap?.moveCamera(
            CameraUpdate.fitBounds(latLngBounds, padding)
                .animate(CameraAnimation.Easing)
        )
    }

    // MARK: - Map Properties

    override fun setMapType(type: NaverMapType) {
        naverMap?.mapType = when (type) {
            NaverMapType.BASIC -> NaverMap.MapType.Basic
            NaverMapType.NAVI -> NaverMap.MapType.Navi
            NaverMapType.SATELLITE -> NaverMap.MapType.Satellite
            NaverMapType.HYBRID -> NaverMap.MapType.Hybrid
            NaverMapType.TERRAIN -> NaverMap.MapType.Terrain
            NaverMapType.NONE -> NaverMap.MapType.None
        }
    }

    override fun setMapPadding(top: Int, left: Int, bottom: Int, right: Int) {
        naverMap?.setContentPadding(left, top, right, bottom)
    }

    override fun setCompassEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isCompassEnabled = enabled
    }

    override fun setScaleBarEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isScaleBarEnabled = enabled
    }

    override fun setZoomControlEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isZoomControlEnabled = enabled
    }

    override fun setLocationButtonEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isLocationButtonEnabled = enabled
    }

    override fun setBuildingHeight(height: Float) {
        naverMap?.buildingHeight = height
    }

    override fun setNightModeEnabled(enabled: Boolean) {
        naverMap?.isNightModeEnabled = enabled
    }

    override fun setMinZoomLevel(level: Double) {
        naverMap?.minZoom = level
    }

    override fun setMaxZoomLevel(level: Double) {
        naverMap?.maxZoom = level
    }

    override fun setScrollGesturesEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isScrollGesturesEnabled = enabled
    }

    override fun setZoomGesturesEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isZoomGesturesEnabled = enabled
    }

    override fun setTiltGesturesEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isTiltGesturesEnabled = enabled
    }

    override fun setRotateGesturesEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isRotateGesturesEnabled = enabled
    }

    override fun setStopGesturesEnabled(enabled: Boolean) {
        naverMap?.uiSettings?.isStopGesturesEnabled = enabled
    }

    override fun setLocationTrackingMode(mode: NaverMapLocationTrackingMode) {
        naverMap?.locationTrackingMode = when (mode) {
            NaverMapLocationTrackingMode.NONE -> LocationTrackingMode.None
            NaverMapLocationTrackingMode.NO_FOLLOW -> LocationTrackingMode.NoFollow
            NaverMapLocationTrackingMode.FOLLOW -> LocationTrackingMode.Follow
            NaverMapLocationTrackingMode.FACE -> LocationTrackingMode.Face
        }
    }

    override fun setLayerGroupEnabled(group: String, enabled: Boolean) {
        val layerGroup = when (group) {
            "building" -> NaverMap.LAYER_GROUP_BUILDING
            "ctt" -> NaverMap.LAYER_GROUP_TRAFFIC
            "transit" -> NaverMap.LAYER_GROUP_TRANSIT
            "bike" -> NaverMap.LAYER_GROUP_BICYCLE
            "mountain" -> NaverMap.LAYER_GROUP_MOUNTAIN
            "landparcel" -> NaverMap.LAYER_GROUP_CADASTRAL
            else -> return
        }
        naverMap?.setLayerGroupEnabled(layerGroup, enabled)
    }

    // MARK: - Markers

    override fun addMarker(data: NaverMapMarkerData) {
        val map = naverMap ?: return

        val marker = Marker().apply {
            position = LatLng(data.coordinate.latitude, data.coordinate.longitude)
            if (data.zIndex != 0) zIndex = data.zIndex
            if (data.rotation != 0f) angle = data.rotation
            isFlat = data.flat
            alpha = data.alpha

            if (data.pinColor != 0 && (data.pinColor ushr 24) != 0) {
                iconTintColor = data.pinColor
            }

            if (data.image.isNotEmpty()) {
                loadMarkerImage(this, data.image)
                if (data.width > 0) width = dpToPx(data.width)
                if (data.height > 0) height = dpToPx(data.height)
            }

            setOnClickListener {
                delegate?.onMarkerClick(data.identifier)
                true
            }

            this.map = map
        }

        markers[data.identifier] = marker
    }

    override fun updateMarker(data: NaverMapMarkerData) {
        markers[data.identifier]?.apply {
            position = LatLng(data.coordinate.latitude, data.coordinate.longitude)
            if (data.width > 0) width = dpToPx(data.width)
            if (data.height > 0) height = dpToPx(data.height)
            zIndex = data.zIndex
            angle = data.rotation
            isFlat = data.flat
            alpha = data.alpha
            if (data.pinColor != 0) iconTintColor = data.pinColor
            if (data.image.isNotEmpty()) loadMarkerImage(this, data.image)
        }
    }

    override fun removeMarker(identifier: String) {
        markers[identifier]?.map = null
        markers.remove(identifier)
    }

    private fun loadMarkerImage(marker: Marker, url: String) {
        thread {
            try {
                val connection = URL(url).openConnection()
                connection.connect()
                val inputStream = connection.getInputStream()
                val bitmap = BitmapFactory.decodeStream(inputStream)
                inputStream.close()

                containerView?.post {
                    marker.icon = OverlayImage.fromBitmap(bitmap)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    // MARK: - Polylines

    private fun lineCapType(type: Int): PolylineOverlay.LineCap {
        return when (type) {
            1 -> PolylineOverlay.LineCap.Round
            2 -> PolylineOverlay.LineCap.Square
            else -> PolylineOverlay.LineCap.Butt
        }
    }

    private fun lineJoinType(type: Int): PolylineOverlay.LineJoin {
        return when (type) {
            1 -> PolylineOverlay.LineJoin.Round
            2 -> PolylineOverlay.LineJoin.Bevel
            else -> PolylineOverlay.LineJoin.Miter
        }
    }

    override fun addPolyline(data: NaverMapPolylineData) {
        val map = naverMap ?: return
        val coords = data.coordinates.map { LatLng(it.latitude, it.longitude) }
        if (coords.size < 2) return

        val polyline = PolylineOverlay().apply {
            this.coords = coords
            width = dpToPx(data.strokeWidth)
            color = data.strokeColor
            zIndex = data.zIndex
            capType = lineCapType(data.lineCap)
            joinType = lineJoinType(data.lineJoin)
            if (data.pattern.isNotEmpty()) {
                setPattern(*data.pattern.toIntArray())
            }
            this.map = map
        }

        polylines[data.identifier] = polyline
    }

    override fun updatePolyline(data: NaverMapPolylineData) {
        val polyline = polylines[data.identifier] ?: return
        val coords = data.coordinates.map { LatLng(it.latitude, it.longitude) }
        if (coords.size >= 2) {
            polyline.coords = coords
        }
        polyline.width = dpToPx(data.strokeWidth)
        polyline.color = data.strokeColor
        polyline.zIndex = data.zIndex
        polyline.capType = lineCapType(data.lineCap)
        polyline.joinType = lineJoinType(data.lineJoin)
        if (data.pattern.isNotEmpty()) {
            polyline.setPattern(*data.pattern.toIntArray())
        }
    }

    override fun removePolyline(identifier: String) {
        polylines[identifier]?.map = null
        polylines.remove(identifier)
    }

    // MARK: - Polygons

    override fun addPolygon(data: NaverMapPolygonData) {
        val map = naverMap ?: return
        val coords = data.coordinates.map { LatLng(it.latitude, it.longitude) }
        if (coords.size < 3) return

        val polygon = PolygonOverlay().apply {
            this.coords = coords
            if (data.holes.isNotEmpty()) {
                this.holes = data.holes.map { hole ->
                    hole.map { LatLng(it.latitude, it.longitude) }
                }
            }
            color = data.fillColor
            outlineColor = data.strokeColor
            outlineWidth = dpToPx(data.strokeWidth)
            zIndex = data.zIndex
            this.map = map
        }

        polygons[data.identifier] = polygon
    }

    override fun updatePolygon(data: NaverMapPolygonData) {
        val polygon = polygons[data.identifier] ?: return
        val coords = data.coordinates.map { LatLng(it.latitude, it.longitude) }
        if (coords.size >= 3) {
            polygon.coords = coords
        }
        if (data.holes.isNotEmpty()) {
            polygon.holes = data.holes.map { hole ->
                hole.map { LatLng(it.latitude, it.longitude) }
            }
        }
        polygon.color = data.fillColor
        polygon.outlineColor = data.strokeColor
        polygon.outlineWidth = dpToPx(data.strokeWidth)
        polygon.zIndex = data.zIndex
    }

    override fun removePolygon(identifier: String) {
        polygons[identifier]?.map = null
        polygons.remove(identifier)
    }

    // MARK: - Circles

    override fun addCircle(data: NaverMapCircleData) {
        val map = naverMap ?: return

        val circle = CircleOverlay().apply {
            center = LatLng(data.center.latitude, data.center.longitude)
            radius = data.radius
            color = data.fillColor
            outlineColor = data.strokeColor
            outlineWidth = dpToPx(data.strokeWidth)
            zIndex = data.zIndex
            this.map = map
        }

        circles[data.identifier] = circle
    }

    override fun updateCircle(data: NaverMapCircleData) {
        val circle = circles[data.identifier] ?: return
        circle.center = LatLng(data.center.latitude, data.center.longitude)
        circle.radius = data.radius
        circle.color = data.fillColor
        circle.outlineColor = data.strokeColor
        circle.outlineWidth = dpToPx(data.strokeWidth)
        circle.zIndex = data.zIndex
    }

    override fun removeCircle(identifier: String) {
        circles[identifier]?.map = null
        circles.remove(identifier)
    }

    // MARK: - Paths

    override fun addPath(data: NaverMapPathData) {
        val map = naverMap ?: return
        val coords = data.coordinates.map { LatLng(it.latitude, it.longitude) }
        if (coords.size < 2) return

        val path = PathOverlay().apply {
            this.coords = coords
            width = dpToPx(data.width)
            outlineWidth = dpToPx(data.outlineWidth)
            color = data.color
            outlineColor = data.outlineColor
            passedColor = data.passedColor
            passedOutlineColor = data.passedOutlineColor
            progress = data.progress.toDouble()
            zIndex = data.zIndex
            if (data.patternInterval > 0) {
                patternInterval = data.patternInterval
            }
            this.map = map
        }

        paths[data.identifier] = path
    }

    override fun updatePath(data: NaverMapPathData) {
        val path = paths[data.identifier] ?: return
        val coords = data.coordinates.map { LatLng(it.latitude, it.longitude) }
        if (coords.size >= 2) {
            path.coords = coords
        }
        path.width = dpToPx(data.width)
        path.outlineWidth = dpToPx(data.outlineWidth)
        path.color = data.color
        path.outlineColor = data.outlineColor
        path.passedColor = data.passedColor
        path.passedOutlineColor = data.passedOutlineColor
        path.progress = data.progress.toDouble()
        path.zIndex = data.zIndex
    }

    override fun removePath(identifier: String) {
        paths[identifier]?.map = null
        paths.remove(identifier)
    }
}
